pkg/reconcileaction/procedures.go

```diff
@@ -0,0 +1,134 @@
+package reconcileaction
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ProcedureList is used to define all the reconcile Procedures for the operator
+type ProcedureList []Procedure
+
+// Newest returns the Procedure with the highest version number
+func (pl ProcedureList) Newest() (*Procedure, error) {
+	if len(pl) < 1 {
+		return nil, fmt.Errorf("empty list of reconcile procedures")
+	}
+
+	// Sort descending by version
+	sort.Slice(pl, func(i, j int) bool {
+		return pl[i].Version() > pl[j].Version()
+	})
+	p := pl[0]
+
+	return &p, nil
+}
+
+// NewestCompatible returns the newest Procedure that is compatible with the
+// provided currentVersion
+func (pl ProcedureList) NewestCompatible(currentVersion int) (*Procedure, error) {
+	if len(pl) < 1 {
+		return nil, errors.New("empty list of reconcile procedures")
+	}
+
+	// Sort descending by version
+	sort.Slice(pl, func(i, j int) bool {
+		return pl[i].Version() > pl[j].Version()
+	})
+
+	// Walk list to find the first that is compatible w/ the currentVersion
+	for _, p := range pl {
+		if p.MinVersion() <= currentVersion {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no procedures are compatible with deployed version %d", currentVersion)
+}
+
+// Procedure defines the complete "procedure" (the set of Action) necessary to
+// completely reconcile the state.
+type Procedure struct {
+	minVersion int
```

Comment:
Instead of minVersion , why dont we define a dict of "compatible versions" ?

Reply:
The "version" here is meant to be just numeric procedure versions, numbered independently of the operator release version (i.e., an operator release may not require a new version at all, or it may introduce multiple depending on upgrade requirements). As such, I really expect the simple `>=` compatibility test to be sufficient for this. Since it's encapsulated from the rest of the code, we could switch it later w/ little difficulty should it become necessary. Sound reasonable?

Reply:
@JohnStrunk hmmm.. compatible version array or dict would sustain better compared to  `>=` , but `>=` could be a start or first version. so, lets keep it as this for now.